src/service/standard-actions-impl.js

```diff
@@ -103,6 +105,41 @@ export class StandardActions {
     );
   }
 
+  /**
+   * Handles initiliazing the theme mode.
+   *
+   * This methode needs to be called on page load to set the `amp-dark-mode`
+   * class on the body if the user prefers the dark mode.
+   */
+  initThemeMode_() {
+    if (this.prefersDarkMode_()) {
+      this.ampdoc.waitForBodyOpen().then((body) => {
+        body.classList.add('amp-dark-mode');
+      });
+    }
+  }
+
+  /**
+   * Checks whether the user prefers dark mode based on local storage and
+   * user's operating systen settings.
+   *
+   * @return {boolean}
+   */
+  prefersDarkMode_() {
+    try {
+      const themeMode = this.ampdoc.win.localStorage.getItem('amp-dark-mode');
+
+      return (
+        'yes' === themeMode ||
+        (!themeMode &&
+          this.ampdoc.win.matchMedia('(prefers-color-scheme: dark)').matches)
+      );
+    } catch (e) {
+      // LocalStorage may not be accessible
+      return this.ampdoc.win.matchMedia('(prefers-color-scheme: dark)').matches;
+    }
```

Comment:
We can clean this up a bit to deduplicate:
```suggestion
      if (themeMode) {
        return 'yes' === themeMode;
      }
    } catch (e) {}
    // LocalStorage may not be accessible
    return this.ampdoc.win.matchMedia('(prefers-color-scheme: dark)').matches;
```

Reply:
@jridgewell Thanks for the suggestion but this would only utilize the `prefers-color-scheme` if the local storage fails. It would be good to have the dark mode class added automatically for first-time visitors and in an incognito mode based on the user’s operating system settings as initially the local storage value would be null.

Reply:
It will use both. If the `localStorage` succeeds and is not `undefined`, it'll test and return that value. If it fails, or is `undefined` (if nothing has been saved before), then it'll use `matchMedia`. The benefit of this approach is we avoid doing the `matchMedia` unless it's necessary.

Reply:
Yes, correct my bad I think I might have misunderstood this earlier. I've updated it.